Polymorphic parameters soundness bug

<!--
Welcome to OCaml's Issue tracker!

OCaml's developers use this tracker for bugs and feature requests only, rather
than user support.

If you have questions about *using* OCaml, please ask at
https://discuss.ocaml.org (more people read Discuss than this tracker, and
you'll get confirmation of whether you've really found a bug or need a new
feature).

If your error came from the OCaml package manager, opam, (messages beginning
`[ERROR] The compilation of ...`), please start at
https://github.com/ocaml/opam-repository/issues/new.

Some libraries and tools which used to be part of OCaml are now maintained
separately. Please post questions about Graphics, Num, camlp4, LablTk, CamlDBM
or OCamlTopWin on Discuss or on their respective issue trackers (see [README.adoc](https://github.com/ocaml/ocaml/blob/trunk/README.adoc#separately-maintained-components)
for a full list).
-->

The following results in a segfault on trunk:
```ocaml
make runtop -j 

OCaml version 5.5.0+dev0-2025-04-28
Enter #help;; for help.

# let should_not_work ([ bad ] : 'a. 'a list) = print_endline bad;;
val should_not_work : ('a. 'a list) -> unit = <fun>
# should_not_work [];;
Segmentation fault: 11
```

I believe the issue lies in `solve_Ppat_constraint` in `typecore.ml` which instantiates any polytype in `expected_ty` (I can make the fix myself, just want to get confirmation that this is indeed the issue)

--- a/Changes
+++ b/Changes
@@ -405,6 +405,10 @@ OCaml 5.5.0
   the marshalled output.
   (David Allsopp, review by Gabriel Scherer)
 
+- #14434, #14652: Protect check_counter_example_pat against polymorphic types,
+  restoring type soundness.
+  (Stephen Dolan and Jacques Garrigue, report and review by Alistair O'Brien)
+
 ### Other libraries:
 
 - #13700, #14454: Use POSIX thread-safe getgrnam_r, getgrgid_r, getpwnam_r,

--- a/testsuite/tests/typing-poly/poly_params.ml
+++ b/testsuite/tests/typing-poly/poly_params.ml
@@ -456,6 +456,12 @@ val g : unit -> ('a. 'a -> 'a) -> unit = <fun>
 let rec f ([] : 'a. 'a list) = ()
 
 [%%expect{|
+Line 1, characters 11-27:
+1 | let rec f ([] : 'a. 'a list) = ()
+               ^^^^^^^^^^^^^^^^
+Warning 8 [partial-match]: this pattern-matching is not exhaustive.
+  Here is an example of a case that is not matched: "_::_"
+
 val f : ('a. 'a list) -> unit = <fun>
 |}]
 
@@ -566,4 +572,35 @@ Error: The syntactic arity of the function doesn't match the type constraint:
           "fun ... gadt_pat -> fun ..."
           where "gadt_pat" is the pattern with the GADT constructor that
           introduces the local type equation on "a".
+|}]
+
+(* Exhaustiveness check works for annotated polymorphic
+   parameters.
+
+   See https://github.com/ocaml/ocaml/issues/14434 *)
+let should_not_be_exhaustive ([ bad ] : 'a. 'a list) = print_endline bad;;
+should_not_be_exhaustive []
+[%%expect{|
+Line 1, characters 30-51:
+1 | let should_not_be_exhaustive ([ bad ] : 'a. 'a list) = print_endline bad;;
+                                  ^^^^^^^^^^^^^^^^^^^^^
+Warning 8 [partial-match]: this pattern-matching is not exhaustive.
+  Here is an example of a case that is not matched: "bad::_::_"
+
+val should_not_be_exhaustive : ('a. 'a list) -> unit = <fun>
+Exception: Match_failure ("", 1, 29).
+|}]
+
+let should_not_be_exhaustive : ('a. 'a list) -> unit =
+  fun [ bad ] -> print_endline bad;;
+should_not_be_exhaustive []
+[%%expect{|
+Line 2, characters 6-13:
+2 |   fun [ bad ] -> print_endline bad;;
+          ^^^^^^^
+Warning 8 [partial-match]: this pattern-matching is not exhaustive.
+  Here is an example of a case that is not matched: "bad::_::_"
+
+val should_not_be_exhaustive : ('a. 'a list) -> unit = <fun>
+Exception: Match_failure ("", 2, 6).
 |}]

--- a/typing/typecore.ml
+++ b/typing/typecore.ml
@@ -2593,6 +2593,7 @@ let enter_nonsplit_or info =
 let rec check_counter_example_pat
     ~info ~(penv : Pattern_env.t) type_pat_state tp expected_ty k =
   assert (penv.in_counterexample = true);
+  assert (not (is_Tpoly expected_ty));
   let check_rec ?(info=info) ?(penv=penv) =
     check_counter_example_pat ~info ~penv type_pat_state in
   let loc = tp.pat_loc in
@@ -2732,6 +2733,9 @@ let check_counter_example_pat ~counter_example_args penv tp expected_ty =
      way -- one of the functions it calls writes an entry into
      [tps_pattern_forces] -- so we can just ignore module patterns. *)
   let type_pat_state = create_type_pat_state Modules_ignored in
+  let expected_ty =
+    with_level ~level:generic_level (fun () -> maybe_instance_poly expected_ty)
+  in
   wrap_trace_gadt_instances ~force:true !!penv
     (check_counter_example_pat ~info:counter_example_args ~penv
        type_pat_state tp expected_ty)

